src/backend/xacheck.go

```diff
@@ -0,0 +1,340 @@
+ *
+ * Copyright 2018 The Radon Authors.
+ * Code is licensed under the GPLv3.
+ *
+ */
+
+package backend
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"config"
+
+	"github.com/pkg/errors"
+	"github.com/xelabs/go-mysqlstack/xlog"
+)
+
+const (
+	xacheckJSONFile = "xacheck.json"
+)
+
+const (
+	xaRedoError = "error"
+	xaRedoOk    = "ok"
+
+	txnXACommitErrStateCommit   = "commit"
+	txnXACommitErrStateRollback = "rollback"
+)
+
+// XaCommitErr tuple.
+type XaCommitErr struct {
+	Time     string   `json:"time"`
+	Xaid     string   `json:"xaid"`
+	State    string   `json:"state"`
+}
+
+// XaCommitErrs tuple
+type XaCommitErrs struct {
+	Logs []*XaCommitErr `json:"xacommit-errs"`
+}
+
+// XaCheck tuple.
+type XaCheck struct {
+	log     *xlog.Log
+	dir     string
+	scatter *Scatter
+	retrys  map[string]*XaCommitErr
+	done    chan bool
+	ticker  *time.Ticker
+	wg      sync.WaitGroup
+	mu      sync.RWMutex
+}
+
+// NewXaCheck creates the XaCheck tuple.
+func NewXaCheck(scatter *Scatter, conf *config.ScatterConfig) *XaCheck {
+	return &XaCheck{
+		log:     scatter.log,
+		dir:     conf.XaCheckDir,
+		scatter: scatter,
+		retrys:  make(map[string]*XaCommitErr),
+		done:    make(chan bool),
+		ticker:  time.NewTicker(time.Duration(time.Second * time.Duration(conf.XaCheckInterval))),
+	}
+}
+
+// Init used to init xa check goroutine.
+func (xc *XaCheck) Init() error {
+	log := xc.log
+
+	// If the xc.dir is already a directory, MkdirAll does nothing
+	// if the dir is one file, return err
+	if err := os.MkdirAll(xc.dir, 0744); err != nil {
+		return err
+	}
+
+	if err := xc.LoadXaCommitErrLogs(); err != nil {
+		return err
+	}
+
+	xc.wg.Add(1)
+	go func(dc *XaCheck) {
+		defer dc.wg.Done()
+		dc.xaCommitCheck()
+	}(xc)
+
+	log.Info("xacheck.init.done")
+	return nil
+}
+
+func (xc *XaCheck) addXaCommitErrLog(new *XaCommitErr) error {
+	log := xc.log
+	log.Info("xc.addXaCommitErrLog.add:+%v", new)
+
+	if _, ok := xc.retrys[new.Xaid]; ok {
+		log.Error("xacheck.addXACommitErrLog.xaid[%v].duplicate", new)
+		return errors.Errorf("xacheck.addXACommitErrLog.xaid[%v].duplicate", new.Xaid)
+	}
+
+	xc.retrys[new.Xaid] = new
+	return nil
+}
+
+// flushXaCommitErrLog is used to write the xaCommitErrlogs to the file.
+func (xc *XaCheck) flushXaCommitErrLog() error {
+	log := xc.log
+	file := path.Join(xc.dir, xacheckJSONFile)
+
+	var xaCommitErrs XaCommitErrs
+	for _, v := range xc.retrys {
+		xaCommitErrs.Logs = append(xaCommitErrs.Logs, v)
+	}
+
+	log.Info("xacheck.flush.to.file[%v].backends.conf:%+v", file, xaCommitErrs)
+	if err := config.WriteConfig(file, xaCommitErrs); err != nil {
+		log.Panicf("xacheck.flush.config.to.file[%v].error:%v", file, err)
+		return err
+	}
+	return nil
+}
+
+// WriteXaCommitErrLog is used to write the xaCommitErrLog into the xacheck file.
+func (xc *XaCheck) WriteXaCommitErrLog(txn *Txn, state string) error {
+	// if the xc is nil, it will be nothing to do
+	if xc == nil {
```

Comment:
The nil check used for? 
From the logic, it can't be nil here, because we shuld call the txnmgr.Init before we use it.

Reply:
sometimes, make coverage will generate the FAIL, I will see:

--- FAIL: TestLongQuery (0.52s)
        Error Trace:    query_test.go:435
	Error:		Expected nil, but got: &sqldb.SQLError{Num:0x451, State:"HY000", Message:"Server maybe lost, please try again", Query:""}

Reply:
Please fixup this testcase, we should not add more codes to fix the testcase.

Reply:
Delete the judgment and  fixup the testcases which related to the commit err and rollback error